File: 30-days-of-code/day05/solution.ex
defmodule Solution do

  def run do
    parse_input()
      |> multiples_of()
      |> Stream.take(10)
      |> Enum.each(&print_multiple/1)
  end

  defp parse_input do
    {n, _} = IO.read(:line) |> Integer.parse
    n
  end

  defp multiples_of(n) do
    Stream.unfold({n, 1}, fn ({n, m}) -> {multiple_of(n, m), {n, m + 1}} end)
  end

  defp multiple_of(n, m) do
    %{:number => n, :multiple => m, :result => n * m}
  end

  defp print_multiple(%{:number => n, :multiple => m, :result => r}) do
    :io.format "~B x ~B = ~B~n", [n, m, r]
  end

end

Solution.run

File: 30-days-of-code/day05/solution.hs
import Control.Applicative
import Control.Monad
import System.IO
import Text.Printf


main :: IO ()
main = do
  n <- getInput
  mapM_ printMultiple $ take 10 $ multiplesOf n 1


getInput :: IO Int
getInput = do
  n_temp <- getLine
  let n = read n_temp :: Int
  return n


multiplesOf :: Int -> Int -> [(Int, Int, Int)]
multiplesOf n multiple =
  let result = n * multiple in
    (n, multiple, result) : multiplesOf n (succ multiple)


printMultiple :: (Int, Int, Int) -> IO ()
printMultiple (n, multiple, result) =
  printf "%d x %d = %d\n" n multiple result
